tot/models.py

```diff
@@ -225,3 +231,302 @@ def maybe_drop_added_dates(self, predicted, df):
+
+    def maybe_drop_added_dates(self, predicted, df):
+        """if Model imputed any dates: removes any dates in predicted which are not in df_test.
+
+        Parameters
+        ----------
+            predicted: pd.DataFrame
+                dataframe containing predicted data
+            df: pd.DataFrame
+                dataframe containing initial data
+
+        Returns
+        -------
+            pd.DataFrame
+                dataframe containing predicted data of dates that are present in df.
+            pd.DataFrame
+                dataframe containing initial data.
+        """
+        (
+            predicted,
+            received_ID_col_pred,
+            received_single_time_series_pred,
+            received_dict_test_pred,
+            _,
+        ) = df_utils.prep_or_copy_df(predicted)
+        df, received_ID_col_df, received_single_time_series_df, received_dict_test_df, _ = df_utils.prep_or_copy_df(df)
+        predicted_new = pd.DataFrame()
+        df_new = pd.DataFrame()
+        for df_name, df_i in df.groupby("ID"):
+            predicted_i = predicted[predicted["ID"] == df_name].copy(deep=True)
+            predicted_i["ds"] = convert_to_datetime(predicted_i["ds"])
+            df_i["ds"] = convert_to_datetime(df_i["ds"])
+            df_i.set_index("ds", inplace=True)
+            predicted_i.set_index("ds", inplace=True)
+            predicted_i = predicted_i.loc[df_i.index]
+            predicted_i = predicted_i.reset_index()
+            df_i = df_i.reset_index()
+            df_new = pd.concat((df_new, df_i), ignore_index=True)
+            predicted_new = pd.concat((predicted_new, predicted_i), ignore_index=True)
+        df = df_utils.return_df_in_original_format(
+            df_new, received_ID_col_df, received_single_time_series_df, received_dict_test_df
+        )
+        predicted = df_utils.return_df_in_original_format(
+            predicted_new, received_ID_col_pred, received_single_time_series_pred, received_dict_test_pred
+        )
+        return predicted, df
+
+    def _predict_raw(self, df):
+        """Computes forecast-origin-wise seasonal naive predictions.
+
+        Predictions are returned in vector format. Predictions are given on a forecast origin basis,
+        not on a target basis.
+
+        Parameters
+        ----------
+            df : pd.DataFrame
+                dataframe containing column ``ds``, ``y``, and optionally``ID`` with all data
+
+        Returns
+        -------
+            pd.Series
+                timestamps referring to the start of the predictions.
+            np.array
+                array containing the predictions
+        """
+        # Receives df with single ID column
+        assert len(df["ID"].unique()) == 1
+
+        dates = df["ds"].iloc[self.K : -self.n_forecasts + 1].reset_index(drop=True)
+        # assemble last values based on K
+        last_k_vals_arrays = [df["y"].iloc[i : i + self.K].values for i in range(0, dates.shape[0])]
```

Comment:
pandas may complain about this syntax:
`df["y"].iloc[i:b]`
or I may remember wrong